src/ui/SimulationControls.tsx

```diff
@@ -0,0 +1,145 @@
+
+export default function VisualWindows() {
+  const coreContext = useContext(CoreContext);
+  // used temporary as for state change signalization
+  const [automatonCore, setAutomatonCore] = useState<IAutomatonCore | null>(null);
+
+  const [simulationWord, setSimulationWord] = useState("aabc");
+  const [simulationWordPos, setSimulationWordPos] = useState(0);
+  const [wordRead, setWordRead] = useState<string[]>([]);
+  const [wordRemaining, setWordRemaining] = useState<string[]>([]);
+  const [steppping, setStepping] = useState(false);
+  const [stepInterval, setStepInterval] = useState<number>(1.0);
+
+  const startSimulation = () => {
+    if (coreContext == undefined) {
+      return;
+    }
+
+    const word = simulationWord.split("");
+
+    const core = coreContext.primary;
+    if (core.kind == Kind.AUTOMATON) {
+      core.runStart(word);
+      setAutomatonCore(core);
+      setSimulationWordPos(0);
+      setWordRemaining(word);
+      setWordRead([]);
+    }
+  };
+  const stopSimulation = () => {
+    if (coreContext == undefined) {
+      return;
+    }
+    const core = coreContext.primary;
+    if (core.kind == Kind.AUTOMATON) {
+      core.runEnd();
+      setAutomatonCore(null);
+      setSimulationWordPos(0);
+      setWordRemaining([]);
+      setWordRead([]);
+      stopTimer();
+    }
+  };
+
+  // returns wether succeded
+  function nextStep(): boolean {
+    if (automatonCore == null) {
+      return false;
+    }
+
+    const error = automatonCore.runNext();
+    if (error) {
+      if (error.details === "Input end reached") {
+        alert("Si na konci slova!");
+      } else {
+        console.log(error);
+      }
+      return false;
+    }
+
+    setWordRead(Array.prototype.concat(wordRead, wordRemaining[0]));
+    setWordRemaining(wordRemaining.slice(1));
+    setSimulationWordPos(simulationWordPos + 1);
+
+    console.log(Array.prototype.concat(wordRead, wordRemaining[0]), wordRemaining.slice(1));
+
+    return true;
+  }
+
+  function prevStep() {
+    if (automatonCore == null) {
+      return false;
+    }
+
+    const error = automatonCore.runUndo();
+    if (error) {
+      if (error.details === "Cannot undo because command history is empty.") {
+        alert("Si na začiatku slova!");
+      } else {
+        console.log(error);
+      }
+      return false;
+    }
+
+    setWordRemaining(Array.prototype.concat(wordRead[0], wordRemaining));
+    setWordRead(wordRead.slice(1));
+    setSimulationWordPos(simulationWordPos - 1);
+
+    return true;
+  }
+
+  const [timeInterval, setTimeInterval] = useState<number | undefined>(undefined);
+
+  // Function to start the timer
+  const startTimer = () => {
+    setStepping(true);
+    setTimeInterval(setInterval(() => {
+      if (!nextStep()) {
+        stopTimer();
+      }
+    }, stepInterval * 1000));
+  };
+
+  // Function to pause the timer
+  const stopTimer = () => {
+    setStepping(false);
+    clearInterval(timeInterval);
+  };
```

Comment:
```suggestion
  const intervalRef = useRef<number | null>(null);

  const startTimer = () => {
    setStepping(true);
    if (intervalRef.current === null) { // prevent multiple intervals
      intervalRef.current = window.setInterval(() => {
        if (!nextStep()) {
          stopTimer();
        }
      }, stepInterval * 1000);
    }
  };

  const stopTimer = () => {
    setStepping(false);
    if (intervalRef.current !== null) {
      clearInterval(intervalRef.current);
      intervalRef.current = null;
    }
  };

  // Cleanup timer on unmount to avoid dangling timers
  useEffect(() => () => {
    if (intervalRef.current !== null) {
      clearInterval(intervalRef.current);
    }
  }, []);
```
pôvodné riešenie nezastaví korektne timer, alert vždy vybehne znova, toto by malo zaručiť správne ukončenie timera